File: EjemploExposicion/SRP.js
// Incorrecto
class Order {
    constructor(items, prices, taxRate) {
        this.items = items;
        this.prices = prices;
        this.taxRate = taxRate;
    }

    calculateTotal() {
        let total = 0;
        for (let price of this.prices) {
            total += price;
        }
        const discount = 5;
        const tax = (total - discount) * this.taxRate;
        return total - discount + tax;
    }

    generateInvoice() {
        const total = this.calculateTotal();
        let invoice = "Factura:\n";
        for (let i = 0; i < this.items.length; i++) {
            invoice += `${this.items[i]}: $${this.prices[i]}\n`;
        }
        invoice += `Total (incluyendo impuestos): $${total}\n`;
        return invoice;
    }

    sendConfirmationEmail(email) {
        const invoice = this.generateInvoice();
        console.log(`Enviando email a ${email} con la siguiente Factura:\n${invoice}`);
    }

    applyDiscount(discount) {
        let total = 0;
        for (let price of this.prices) {
            total += price;
        }
        total -= discount;
        return total > 0 ? total : 0;
    }
}

// Ejemplo de uso
class Order {
    constructor(items) {
        this.items = items;
    }

    getSubtotal() {
        return this.items.reduce((subtotal, [item, price]) => subtotal + price, 0);
    }
}

class TaxCalculator {
    calculateTax(subtotal, taxRate) {
        return subtotal * taxRate;
    }
}

class DiscountManager {
    applyDiscount(subtotal, discount) {
        const total = subtotal - discount;
        return total > 0 ? total : 0;
    }
}

class InvoiceGenerator {
    generateInvoice(order, subtotal, tax, total) {
        let invoice = "Factura:\n";
        for (const [item, price] of order.items) {
            invoice += `${item}: $${price}\n`;
        }
        invoice += `Subtotal: $${subtotal}\n`;
        invoice += `Impuesto: $${tax}\n`;
        invoice += `Total (incluyendo impuestos): $${total}\n`;
        return invoice;
    }
}

class EmailSender {
    sendEmail(email, content) {
        console.log(`Enviando email a ${email} con el siguiente contenido:\n${content}`);
    }
}

// Ejemplo de uso
const order = new Order([["Libro", 12.99], ["Lapiz", 0.99]]);
const taxCalculator = new TaxCalculator();
const discountManager = new DiscountManager();
const invoiceGenerator = new InvoiceGenerator();
const emailSender = new EmailSender();

const subtotal = order.getSubtotal();
const discountedTotal = discountManager.applyDiscount(subtotal, 5);
const tax = taxCalculator.calculateTax(discountedTotal, 0.1);
const finalTotal = discountedTotal + tax;

const invoice = invoiceGenerator.generateInvoice(order, subtotal, tax, finalTotal);
emailSender.sendEmail("comprador@example.com", invoice);

File: EjemploExposicion/SRP.py
# Incorrecto
class Order:
    def __init__(self, Items, TaxRate):
        self.Items = Items 
        self.TaxRate = TaxRate
    
    def CalculateTotal(self):
        Total = sum(Price for Item, Price in self.Items)
        Discount = 5  
        Tax = (Total - Discount) * self.TaxRate
        return Total - Discount + Tax
    
    def GenerateInvoice(self):
        Total = self.CalculateTotal()
        Invoice = "Factura:\n"
        for Item, Price in self.Items:
            Invoice += f"{Item}: ${Price}\n"
        Invoice += f"Total (incluyendo impuestos): ${Total}\n"
        return Invoice
    
    def SendConfirmationEmail(self, Email):
        Invoice = self.GenerateInvoice()
        print(f"Enviando email a {Email} con la siguiente Factura:\n{Invoice}")
    
    def ApplyDiscount(self, Discount):
        Total = sum(Price for Item, Price in self.Items) - Discount
        return Total if Total > 0 else 0 


OrderInstance = Order([("Libro", 12.99), ("Lapicero", 0.99)], 0.1)
print(OrderInstance.CalculateTotal())
print(OrderInstance.GenerateInvoice())
OrderInstance.SendConfirmationEmail("customer@example.com")


# Correcto
class Order:
    def __init__(self, Items):
        self.Items = Items
    
    def GetSubtotal(self):
        return sum(Price for _, Price in self.Items)

class TaxCalculator:
    def CalculateTax(self, Subtotal, TaxRate):
        return Subtotal * TaxRate

class DiscountManager:
    def ApplyDiscount(self, Subtotal, Discount):
        Total = Subtotal - Discount
        return Total if Total > 0 else 0

class InvoiceGenerator:
    def GenerateInvoice(self, Order, Subtotal, Tax, Total):
        Invoice = "Factura:\n"
        for Item, Price in Order.Items:
            Invoice += f"{Item}: ${Price}\n"
        Invoice += f"Subtotal: ${Subtotal}\n"
        Invoice += f"Impuesto: ${Tax}\n"
        Invoice += f"Total (incluyendo impuestos): ${Total}\n"
        return Invoice

class EmailSender:
    def SendEmail(self, Email, Content):
        print(f"Enviando email a {Email} con el siguiente contenido:\n{Content}")


OrderInstance = Order([("Libro", 12.99), ("Lapiz", 0.99)])
TaxCalculatorInstance = TaxCalculator()
DiscountManagerInstance = DiscountManager()
InvoiceGeneratorInstance = InvoiceGenerator()
EmailSenderInstance = EmailSender()

Subtotal = OrderInstance.GetSubtotal()
DiscountedTotal = DiscountManagerInstance.ApplyDiscount(Subtotal, 5)
Tax = TaxCalculatorInstance.CalculateTax(DiscountedTotal, 0.1)
FinalTotal = DiscountedTotal + Tax

Factura = InvoiceGeneratorInstance.GenerateInvoice(OrderInstance, Subtotal, Tax, FinalTotal)
EmailSenderInstance.SendEmail("comprador@example.com", Factura)
